### PR TITLE
Override modulo zero to equal zero

### PR DESCRIPTION
**Pull request by mlutze**
Fixes https://github.com/flix/flix/issues/1967

### Features
* `x % 0 == 0`
  * [x] impl
  * [x] test

### General requirements
* [x] Don't make a complete mess
* [x] Don't break anything else
* [x] Provide documentation
* [x] Add license to new files

### CODE REVIEW

**Comment by mlutze** [comment]
Before merging I'd like to see if I can get rid some of the duplication (I copy-pasted code from the division changes)

**Comment by magnus-madsen** [comment]
Does this belong in the bytecode or should we just rewrite earlier?

**Comment by mlutze** [comment]
Good point. Rewriting earlier would certainly be more readable than the byte code stuff.

**Comment by mlutze** [comment]
Do you have an idea of which earlier phase should handle this?

**Comment by magnus-madsen** [comment]
> 
> 
> Do you have an idea of which earlier phase should handle this?

Good question. I was wondering about it too. I am thinking the `Weeder` because its the earliest phase? If we put it in a middle phase it might be very confusing.

**Comment by mlutze** [comment]
I don't think we can put it before `Monomorph` because `/` comes from a type class, so the `check-equal-to-zero` operation isn't defined for every usage of `/`. (Unless we add it to the type class, but that seems a little extreme.)

**Comment by magnus-madsen** [comment]
> 
> 
> I don't think we can put it before `Monomorph` because `/` comes from a type class, so the `check-equal-to-zero` operation isn't defined for every usage of `/`. (Unless we add it to the type class, but that seems a little extreme.)

Check `Weeder` there are some intrisics called `INT32_DIV` and so forth!

**Comment by mlutze** [comment]
```
e1 $INT32_DIV$ e2
```
becomes
```
let x = e1;
let y = e2;
if (y $INT32_EQ$ 0) {
  0
} else {
  x / y
}
```

Is there already a precedent for adding "synthetic identifiers"?

**Comment by magnus-madsen** [comment]
> Is there already a precedent for adding "synthetic identifiers"?

They should start with an underscore and are usually called `tmp` or some such. You have to be careful not to construct shadowed variables (that would be reported by the Redundancy phase).

**Comment by magnus-madsen** [comment]
Question: Could this simply be done in Flix code itself? Inside the instances for div and mod?

**Comment by mlutze** [comment]
> Question: Could this simply be done in Flix code itself? Inside the instances for div and mod?

Yes, it could. I think the downside of this is that it's no longer Flix which has this "total division" property, but rather the Flix (min) library. But it's worth it IMO since this would be a much simpler solution overall.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 1]
```diff
@@ -1368,72 +1364,6 @@ object GenExpression {
...
-          visitor.visitFieldInsn(GETSTATIC, JvmName.BigInteger.toInternalName, "ZERO",
-            JvmType.BigInteger.toDescriptor)
-          visitor.visitJumpInsn(GOTO, endDiv)
-          visitor.visitLabel(div)
-          visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.BigInteger.toInternalName, "divide",
-            AsmOps.getMethodDescriptor(List(JvmType.BigInteger), JvmType.BigInteger), false)
-          visitor.visitLabel(endDiv)
-        case _ => InternalCompilerException(s"Unexpected sematic operator: $sop.")
-      }
     } else {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala:1367]: Feels amazing to remove so much code :)

**Pull request merged**